agent/consul/controller/controller.go

```diff
@@ -0,0 +1,258 @@
+		baseBackoff: 5 * time.Millisecond,
+		maxBackoff:  1000 * time.Second,
+		makeQueue:   RunWorkQueue,
+	}
+}
+
+// Subscribe tells the controller to subscribe to updates for config entries of the
+// given kind and with the associated enterprise metadata. This should only ever be
+// called prior to running Start.
+func (c *controller) Subscribe(request *stream.SubscribeRequest, transformers ...func(entry structs.ConfigEntry) []Request) Controller {
+	c.subscriptions = append(c.subscriptions, subscription{
+		request:      request,
+		transformers: transformers,
+	})
+	return c
+}
+
+// WithBackoff changes the base and maximum backoff values for the Controller's
+// Request retry rate limiter. This should only ever be called prior to
+// running Start.
+func (c *controller) WithBackoff(base, max time.Duration) Controller {
+	c.baseBackoff = base
+	c.maxBackoff = max
+	return c
+}
+
+// WithWorkers sets the number of worker goroutines used to process the queue
+// this defaults to 1 goroutine.
+func (c *controller) WithWorkers(i int) Controller {
+	if i <= 0 {
+		i = 1
+	}
+	c.workers = i
+	return c
+}
+
+// WithQueueFactory changes the initialization method for the Controller's work
+// queue, this is predominantly just used for testing. This should only ever be called
+// prior to running Start.
+func (c *controller) WithQueueFactory(fn func(ctx context.Context, baseBackoff time.Duration, maxBackoff time.Duration) WorkQueue) Controller {
+	c.makeQueue = fn
+	return c
+}
+
+// Start begins the Controller's main processing loop. When the given
+// context is canceled, the Controller stops processing any remaining work.
+// The Start function should only ever be called once.
+func (c *controller) Start(ctx context.Context) error {
+	group, groupCtx := errgroup.WithContext(ctx)
+
+	// set up our queue
+	c.work = c.makeQueue(groupCtx, c.baseBackoff, c.maxBackoff)
+
+	for _, sub := range c.subscriptions {
+		// store a reference for the closure
+		sub := sub
+		group.Go(func() error {
+			var index uint64
+
+			subscription, err := c.publisher.Subscribe(sub.request)
+			if err != nil {
+				return err
+			}
+			defer subscription.Unsubscribe()
+
+			for {
+				event, err := subscription.Next(ctx)
+				switch {
+				case errors.Is(err, context.Canceled):
+					return nil
+				case err != nil:
+					return err
+				}
+
+				if event.IsFramingEvent() {
+					continue
+				}
+
+				if event.Index <= index {
+					continue
+				}
+
+				index = event.Index
+
+				if err := c.processEvent(sub, event); err != nil {
+					return err
+				}
+			}
+		})
+	}
+
+	for i := 0; i < c.workers; i++ {
+		group.Go(func() error {
+			for {
+				request, shutdown := c.work.Get()
+				if shutdown {
+					// Stop working
+					return nil
+				}
+				c.reconcileHandler(groupCtx, request)
+				// Done is called here because it is required to be called
+				// when we've finished processing each request
+				c.work.Done(request)
+			}
+		})
+	}
+
+	<-groupCtx.Done()
```

Comment:
#nit I find it odd that a func named `Start()` is blocking. I'd name it `Run()` or make it non-blocking

Reply:
👍  I'm fine with renaming `Run`, that's used elsewhere in our structs that are meant to have entrypoints invoked in long-running goroutines.